File: go/twelve-days/twelve_days.go
package twelve

import "strings"

type Day struct {
    ordinal, gift string
}

var carol = map[int]*Day{
    1:  {ordinal: "first",    gift: "a Partridge in a Pear Tree"},
    2:  {ordinal: "second",   gift: "two Turtle Doves"},
    3:  {ordinal: "third",    gift: "three French Hens"},
    4:  {ordinal: "fourth",   gift: "four Calling Birds"},
    5:  {ordinal: "fifth",    gift: "five Gold Rings"},
    6:  {ordinal: "sixth",    gift: "six Geese-a-Laying"},
    7:  {ordinal: "seventh",  gift: "seven Swans-a-Swimming"},
    8:  {ordinal: "eighth",   gift: "eight Maids-a-Milking"},
    9:  {ordinal: "ninth",    gift: "nine Ladies Dancing"},
    10: {ordinal: "tenth",    gift: "ten Lords-a-Leaping"},
    11: {ordinal: "eleventh", gift: "eleven Pipers Piping"},
    12: {ordinal: "twelfth",  gift: "twelve Drummers Drumming"},
}

func Verse(i int) string {
    var sb strings.Builder
    sb.WriteString("On the ")
    sb.WriteString(carol[i].ordinal)
    sb.WriteString(" day of Christmas my true love gave to me: ")
    for c := i; c > 0; c-- {
        switch {
            case i == 1:
                sb.WriteString(carol[c].gift)
                sb.WriteString(".")
            case c == 1:
                sb.WriteString("and ")
                sb.WriteString(carol[c].gift)
                sb.WriteString(".")
            default:
                sb.WriteString(carol[c].gift)
                sb.WriteString(", ")
        }
    }
    return sb.String()
}

func Song() string {
    var sb strings.Builder
    for i := 1; i <= 12; i++ {
        sb.WriteString(Verse(i))
        if i != 12 {
            sb.WriteString("\n")
        }
    }
    return sb.String()
}

File: python/twelve-days/twelve_days.py
class Verse:

    def __init__(self, ord, gift):
        self.ord = ord
        self.gift = gift


carol = [
    Verse("first", "a Partridge in a Pear Tree"),
    Verse("second", "two Turtle Doves"),
    Verse("third", "three French Hens"),
    Verse("fourth", "four Calling Birds"),
    Verse("fifth", "five Gold Rings"),
    Verse("sixth", "six Geese-a-Laying"),
    Verse("seventh", "seven Swans-a-Swimming"),
    Verse("eighth", "eight Maids-a-Milking"),
    Verse("ninth", "nine Ladies Dancing"),
    Verse("tenth", "ten Lords-a-Leaping"),
    Verse("eleventh", "eleven Pipers Piping"),
    Verse("twelfth", "twelve Drummers Drumming"),
]


def recite(start_verse, end_verse: int) -> str:
    return [verse(i) for i in range(start_verse, end_verse + 1)]


def verse(n: int) -> str:
    result = [
        f"On the {carol[n - 1].ord} day of Christmas my true love gave to me: "
    ]
    c = n
    while c:
        if n == 1:
            result.append(f"{carol[c - 1].gift}.")
        elif c == 1:
            result.append(f"and {carol[c - 1].gift}.")
        else:
            result.append(f"{carol[c - 1].gift}, ")
        c -= 1
    return "".join(result)
